runtime/src/accounts_index.rs

```diff
@@ -194,15 +196,70 @@ impl<T: 'static + Clone> AccountsIndex<T> {
         AccountsIndexIterator::new(&self.account_maps, range)
     }
 
-    fn do_scan_accounts<'a, F, R>(&'a self, ancestors: &Ancestors, mut func: F, range: Option<R>)
-    where
+    fn do_checked_scan_accounts<'a, F, R>(
+        &'a self,
+        ancestors: &Ancestors,
+        func: F,
+        range: Option<R>,
+    ) where
+        F: FnMut(&Pubkey, (&T, Slot)),
+        R: RangeBounds<Pubkey>,
+    {
+        let max_root = {
+            let mut w_ongoing_scan_roots = self
+                // This lock is also grabbed by clean_accounts(), so clean
+                // has at most cleaned up to the current `max_root` (since
+                // clean only happens *after* BankForks::set_root() which sets
+                // the `max_root`)
+                .ongoing_scan_roots
+                .write()
+                .unwrap();
+            // `max_root()` grabs a lock while
+            // the `ongoing_scan_roots` lock is held,
+            // make sure inverse doesn't happen to avoid
+            // deadlock
+            let max_root = self.max_root();
+            *w_ongoing_scan_roots.entry(max_root).or_default() += 1;
+            max_root
+        };
+
+        self.do_scan_accounts(ancestors, func, range, Some(max_root));
```

Comment:
Changing `Some(max_root) -> None`  should cause `test_store_scan_consistency` to fail